src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java

```diff
@@ -137,78 +142,96 @@ public RepositoryDirectoryValue.Builder fetch(
       return fetchInternal(rule, outputDirectory, directories, env, recordedInputValues, key);
     }
     var state = env.getState(RepoFetchingSkyKeyComputeState::new);
-    var workerFuture = state.workerFuture;
-    if (workerFuture == null) {
+    Environment workerEnv = new RepoFetchingWorkerSkyFunctionEnvironment(state, env);
+    var workerThread = state.workerThread;
+    if (workerThread == null) {
       // No worker is running yet, which means we're just starting to fetch this repo. Start with a
       // clean slate, and create the worker.
       setupRepoRoot(outputDirectory);
-      Environment workerEnv = new RepoFetchingWorkerSkyFunctionEnvironment(state, env);
-      workerFuture =
-          workerExecutorService.submit(
-              () -> {
-                try {
-                  return fetchInternal(
-                      rule,
-                      outputDirectory,
-                      directories,
-                      workerEnv,
-                      state.recordedInputValues,
-                      key);
-                } finally {
-                  state.signalQueue.put(Signal.DONE);
-                }
-              });
-      state.workerFuture = workerFuture;
-    } else {
-      // A worker is already running. This can only mean one thing -- we just had a Skyframe
-      // restart, and need to send over a fresh Environment.
-      state.delegateEnvQueue.put(env);
+      workerThread =
+          Thread.ofVirtual()
+              .name("starlark-repository-" + rule.getName())
+              .start(
+                  () -> {
+                    try {
+                      SkyFunction.Environment workerEnvInThread = state.rendezvousFromWorker(null);
```

Comment:
I'm not a big fan of the use of `Exchanger` here.

1. Often the information exchange is one-way only (here and on line 166). You might as well be using SynchronousQueue.
2. There is no type safety; untyped Objects are being passed around. This is mitigated by the wrapper `rendezvous*` methods but at that point why don't we just use two SynchronousQueues?
3. Following on from the two points above: it's really hard to tell what's going on at any given point. Should we `exchange` by giving something? And/or should we ask for something back? _What_ are we giving or asking back? These questions can be answered, but you need to be looking *really* carefully at the code on both sides to even begin answer them. Whereas previously the "state" was very easy to tell: either the worker sends back a RESTART and expects an Environment back, or it sends back a DONE and is finished.

Reply:
ah sorry I missed your comment explaining your rationale for `Exchanger` before. I now understand the reason better, but still think point 3 above stands. Maybe it's just a matter of naming? Or maybe clearer demarcating of states.

Reply:
Sigh. No, I'm not a fan of `Exchanger`, either and its appearance is an obvious sign of desperation, but I have no better ideas. Trimmed to the bone, the worker thread does this:

```java

try {
  while (!done) {
    var env = getNewEnvironment();
    doThings(env);
    send("success");
  }
} catch (Throwable e) {
  send("failure");
}
```

and the Skyframe function does this:

```java
try 
  send(env);
  var result = receive();
  return result == "restart" ? null : result;
} catch (Throwable e) {
  return "failure";
}
```

and the discrepancy between the two halves is that the worker thread can consider sending a new request and getting a new environment an atomic operation, but the Skyframe thread cannot, because it needs to do work in between. 

It is also the case that when an interrupt happens, the worker thread will end up at `send("failure")`, regardless of whether the Skyframe thread expects it to send or receive a value.  IOW, the Skyframe thread must be ready to receive a failure even when the worker thread signaled that it wants to received something instead. Which is why two blocking data structures inherently doesn't work.

The way I got here is that I thought that it would be easier to model this with two state machines passing messages between each other, then realized that writing a state machine and handling exceptions in parallel isn't nice (I tried), then I realized that executing code is just a state machine whose state is on the stack (and registers, etc.), then I started looking for a bidirectional communication mechanism between two threads, then I found Exchanger.

No, I'm not happy about this but this is about the best I could come up with :(